applications/madini_tracechain.md

```diff
@@ -60,6 +61,22 @@ TraceChain interfaces with both block-chain and a server with a database.Madini
 	•VIP 191
 
```

Comment:
![tracechainweb](https://user-images.githubusercontent.com/8124128/140721833-03579233-edda-4d8f-b4e0-d14b246f7880.jpg)

**Madini TraceChain Technical and Architecture**

TraceChain interfaces with both block-chain and a server with a database.Madini TraceChain relies on the Vechain blockchain a distributed block-chain service.Verified data is broadcasted to the blockchain and stored as hex data. Madini TraceChain relies on the Vechain blockchain a distributed block-chain service.Verified data is broadcasted to the blockchain and stored as hex data also stores the raw data to the SQLite 

Madini TraceChain Technical Details:

•Python 3.8 (Flask)
•VPS / Server.
•Sqlite Database.
•VET Wallet address with VTHO.
•VIP 190

TraceChain consists of Web App, API’s .The Web application is made using JQUERY,JAVASCRIPT and python 3.8 flask which integrates VIP 191 which help send data to the Vechain blockchain as well as saving data to the database.
How does VIP-191 Work?
VIP-191 comes with fee delegation protocol on VeChain.VIP-191 works by delegating a secondary address to pay the transaction cost. When processing, the transaction will be signed by the secondary address, before being sent to the blockchain.VIP-191 is using a specially formed signature composed by the user and the sponsor together.
Just as the picture below has depicted, once the transaction is forged, the user extracts the hash (information digest) and signs it as normal. At the same time, the sponsor also signs the hash but with additional information “for whom”. Then we concatenate the two signatures together, then send the final composed signature along with the transaction itself, to the VeChain network.

Sample Code using VIP 191 To send a transaction to Vechain Blockchain.

@auth.route('/createnew', methods=['GET', 'POST'])
def CreatenewRouted():
if request.method == "POST":
hashdata = str(request.json['hashdata'])
des = str(request.json['des'])
sku = str(request.json['sku'])
pname = str(request.json['pname'])
ptype = str(request.json['type'])
mainimage = str(request.json['mainimage'])
_node_url = node
BlockInfos = requests.get(_node_url + '/blocks/best') 
_BlockRef = BlockInfos.json()['id'][0:18]
_Nonce = randint(10000000, 99999999)
delegated_body = {
"chainTag": chaintag,
"blockRef": _BlockRef,
"expiration": 720,
"clauses": [
{
"to": '',#enter to address
"value": 0,
"data": '0x' +username
}
],
"gasPriceCoef": 0,
"gas": 100000,
"dependsOn": None,
"nonce": _Nonce,
"reserved": {
"features": 1
}
}
print(hashdata)
delegated_tx = transaction.Transaction(delegated_body)
assert delegated_tx.is_delegated() == True
# Sender address
addr_1 = ''
priv_1 = bytes.fromhex('') # Sender private key address
# Gas 5
addr_2 = '' # Gas Payer
priv_2 = bytes.fromhex('')#Gas payer key
h = delegated_tx.get_signing_hash() # Sender hash to be signed.
dh = delegated_tx.get_signing_hash(addr_1) # Gas Payer hash to be signed.
# Sender sign the hash.
# Gas payer sign the hash.
# Concat two parts to forge a legal signature.
sig = cry.secp256k1.sign(h, priv_1) + cry.secp256k1.sign(dh, priv_2)
delegated_tx.set_signature(sig)
ts = datetime.datetime.now()
tx_headers = {'Content-Type': 'application/json', 'accept': 'application/json'}
send_transaction = requests.post(_node_url + '/transactions', json=tx_data, headers=tx_headers)
print('Response from Server: ' + str(send_transaction.content))
prt=str(send_transaction.content)
encoded_bytes = delegated_tx.encode()
new_verif = Verif(bhash=des, mid=username, thash=delegated_tx.get_id(), sku=sku, pname=pname ,mainimage=mainimage ,datetime=ts) #
db.session.add(new_verif)
db.session.commit()
return "<a target='blank' href='https://explore.vechain.org/transactions/"+delegated_tx.get_id()+"'>Explorer</a>"
encoded_bytes = delegated_tx.encode()
tx_data = {'raw': '0x' + encoded_bytes.hex()}
print(tx_data)
ts = datetime.datetime.now()
tx_headers = {'Content-Type': 'application/json', 'accept': 'application/json'}
send_transaction = requests.post(_node_url + '/transactions', json=tx_data, headers=tx_headers)
print('Response from Server: ' + str(send_transaction.content))
prt=str(send_transaction.content)
encoded_bytes = delegated_tx.encode()
new_verif = Verif(bhash=des, mid=username, thash=delegated_tx.get_id(), sku=sku, pname=pname ,mainimage=mainimage ,datetime=ts,ptype=ptype,shape=shape,color=color,carats=carats,clarity=clarity,cut=cut,category=category) #
db.session.add(new_verif)
db.session.commit()
return "<a target='blank' href='https://explore.vechain.org/transactions/"+delegated_tx.get_id()+"'>Explorer</a>"